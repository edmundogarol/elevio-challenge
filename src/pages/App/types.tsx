import { Article } from "../../reducers/App/types";

export type StateProps = {
  searching: boolean;
  articles: Article[];
};

export type DispatchProps = {
  searchTextCall: (text: string) => void;
  searchLog: (searchText: string) => void;
  articleViewLog: (article: number) => void;
};
