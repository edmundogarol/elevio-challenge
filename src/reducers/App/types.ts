export type Article = {
  id: number;
  title: string;
};

export interface AppState {
  articles: Article[];
  searching: boolean;
}
