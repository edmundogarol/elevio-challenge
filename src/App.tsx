import React from "react";
import { Provider } from "react-redux";
import createSagaMiddleware from "redux-saga";
import { createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import { Router, Route } from "react-router-dom";
import { createBrowserHistory } from "history";

import "antd/dist/antd.css";

import AppPage from "./pages/App";
import Article from "./pages/Article";
import appSaga from "./sagas/App";
import appReducer from "./reducers/App";

/**
 * Set up and run app sagas listening for events
 */

const sagaMiddleware = createSagaMiddleware();
const store = createStore(
  appReducer,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);
sagaMiddleware.run(appSaga);

function App() {
  const history = createBrowserHistory();

  return (
    <Provider store={store}>
      <Router history={history}>
        <Route exact path="/" component={AppPage} />
        <Route path="/article/:id/" component={Article} />
      </Router>
    </Provider>
  );
}

export default App;
