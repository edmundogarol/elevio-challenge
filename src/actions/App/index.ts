export const SEARCH_TEXT = "app/SEARCH_TEXT";
export const SEARCHING_TEXT = "app/SEARCHING_TEXT";
export const UPDATE_SEARCHED_ARTICLES = "app/UPDATE_SEARCHED_ARTICLES";

export const SEARCH_LOG = "app/SEARCH_LOG";
export const ARTICLE_VIEW_LOG = "app/ARTICLE_VIEW_LOG";

export const searchText = (text: string) => ({
  type: SEARCH_TEXT,
  payload: {
    text,
  },
});

export const searching = () => ({
  type: SEARCHING_TEXT,
});

export const updateArticles = (articles: any[]) => ({
  type: UPDATE_SEARCHED_ARTICLES,
  payload: {
    articles,
  },
});

export const searchLog = (searchText: string) => ({
  type: SEARCH_LOG,
  payload: {
    searchText,
  },
});

export const articleViewLog = (article: number) => ({
  type: ARTICLE_VIEW_LOG,
  payload: {
    article,
  },
});
