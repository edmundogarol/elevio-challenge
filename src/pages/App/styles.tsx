import styled from "styled-components";

export const AppContainer = styled.div`
  height: 45em;
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
`;

export const Logo = styled.img`
  width: 15em;
`;

export const SearchInputButtonContainer = styled.div`
  display: flex;
  flex-direction: row;
  margin-bottom: 1em;
`;

export const SearchContainer = styled.div`
  width: 80%;
  margin-left: auto;
  margin-right: auto;
  padding: 7em;
  border-radius: 3em;
  display: flex;
  flex-direction: column;
  align-items: center;

  h2 {
    color: #676767;
    margin: unset;
    margin-bottom: 1em;
  }

  input,
  button:not(.search-icon, .ant-alert-close-icon) {
    width: 20em;
    padding: 1em;
    font-size: 1em;
    border-radius: 0.5em;
    border: 1px solid #a7a7a7;
  }
`;
