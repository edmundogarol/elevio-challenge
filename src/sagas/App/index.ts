import { call, all, takeLatest, put } from "redux-saga/effects";
import {
  searching,
  updateArticles,
  SEARCH_TEXT,
  SEARCH_LOG,
  ARTICLE_VIEW_LOG,
} from "../../actions/App";

export async function elevioFetch(url: string, params: any = {}) {
  const apiEndpoint = "https://api.elevio-staging.com";
  const composedURL = new URL(url, apiEndpoint);

  params.headers = {
    ...params.headers,
    "x-api-key": "4a453cfa4c2018b15ee575c8cea281d1",
  };

  params.body = JSON.stringify(params.body);

  try {
    const response = await fetch(composedURL.toString(), params).catch(
      (exc) => {
        console.error(exc);
        throw exc;
      }
    );

    if (response.ok) {
      const data = await response.json();

      return { details: response, data };
    }
  } catch (exc) {
    console.error("Elevio Fetch Error", exc);
    return { details: { ok: false } };
  }
}

export function* searchTextApiCall(action: {
  type: string;
  payload: { text: string };
}) {
  yield put(searching());
  const response: { details: Response; data: any } = yield call(
    elevioFetch,
    `/v1/search/en?query=${action.payload.text}`
  );

  if (response.details.ok) {
    yield put(updateArticles(response.data.results));
  } else {
    console.error("Controller API error", response);
  }
}

export function* searchLogApiCall(action: {
  type: string;
  payload: { search: string };
}) {
  const response: { details: Response; data: any } = yield call(
    elevioFetch,
    `/log-search/?search=${action.payload.search}`
  );

  if (response.details.ok) {
  } else {
    console.error("Search Log API error", response);
  }
}

export function* articleViewLogApiCall(action: {
  type: string;
  payload: { article: number };
}) {
  const response: { details: Response; data: any } = yield call(
    elevioFetch,
    `/article-view-log/?article=${action.payload.article}`
  );

  if (response.details.ok) {
  } else {
    console.error("Article View Log API error", response);
  }
}

export default function* appSaga() {
  yield all([
    takeLatest(SEARCH_TEXT, searchTextApiCall),
    takeLatest(SEARCH_LOG, searchLogApiCall),
    takeLatest(ARTICLE_VIEW_LOG, articleViewLogApiCall),
  ]);
}
