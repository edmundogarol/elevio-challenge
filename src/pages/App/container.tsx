import { connect } from "react-redux";
import { AnyAction, Dispatch } from "redux";

import Ui from "./ui";

import { DispatchProps, StateProps } from "./types";
import { articleViewLog, searchLog, searchText } from "../../actions/App";
import { AppState } from "../../reducers/App/types";

const stateToProps = (state: StateProps) => {
  const { articles, searching } = state;

  return {
    articles,
    searching,
  };
};

const dispatchToProps = (dispatch: Dispatch<AnyAction>) => {
  return {
    searchTextCall: (text: string) => dispatch(searchText(text)),
    searchLog: (searchText: string) => dispatch(searchLog(searchText)),
    articleViewLog: (article: number) => dispatch(articleViewLog(article)),
  };
};

export default connect<StateProps, DispatchProps, {}, AppState>(
  stateToProps,
  dispatchToProps
)(Ui);
