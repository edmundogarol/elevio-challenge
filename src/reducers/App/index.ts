import { SEARCHING_TEXT, UPDATE_SEARCHED_ARTICLES } from "../../actions/App";
import { AppState, Article } from "./types";

const INITIAL_STATE: AppState = {
  articles: [],
  searching: false,
};

export const appReducer = (
  state = INITIAL_STATE,
  action: { type: string; payload: any }
) => {
  switch (action.type) {
    case SEARCHING_TEXT:
      return {
        ...state,
        searching: true,
      };

    case UPDATE_SEARCHED_ARTICLES:
      const { articles } = action.payload;
      return {
        ...state,
        articles: articles as Article[],
        searching: false,
      };

    default:
      return state;
  }
};

export default appReducer;
