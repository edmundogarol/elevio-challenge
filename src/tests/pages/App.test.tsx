import * as React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallowToJson } from "enzyme-to-json";
import { shallow, configure } from "enzyme";

import Ui from "../../pages/App/ui";
import { DispatchProps, StateProps } from "../../pages/App/types";

configure({ adapter: new Adapter() });

describe("Testing App UI", () => {
  const props: StateProps & DispatchProps = {
    searching: false,
    articles: [],
    searchTextCall: () => {},
    searchLog: () => {},
    articleViewLog: () => {},
  };

  let ui = <Ui {...props} />;
  let wrapper = shallow(ui);

  it("check if component renders", () => {
    expect(wrapper.length).toEqual(1);
  });

  it("check rendered ui to match snapshot", () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
