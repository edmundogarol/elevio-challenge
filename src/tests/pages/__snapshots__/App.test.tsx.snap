// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App UI check rendered ui to match snapshot 1`] = `
<styled.div>
  <styled.div>
    <NavLink
      to="/"
    >
      <styled.img
        alt="Elevio"
      />
    </NavLink>
    <h2>
      Search Elevio Articles
    </h2>
    <styled.div>
      <Input
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Search Articles"
        suffix={
          <React.Fragment>
            <Tooltip
              arrowPointAtCenter={false}
              autoAdjustOverflow={true}
              mouseEnterDelay={0.1}
              mouseLeaveDelay={0.1}
              placement="top"
              title="Clear Search"
            >
              <button
                className="search-icon"
                onClick={[Function]}
              >
                <ForwardRef(CloseOutlined) />
              </button>
            </Tooltip>
            <Tooltip
              arrowPointAtCenter={false}
              autoAdjustOverflow={true}
              mouseEnterDelay={0.1}
              mouseLeaveDelay={0.1}
              placement="top"
              title="Search"
            >
              <button
                className="search-icon"
                onClick={[Function]}
              >
                <ForwardRef(SearchOutlined) />
              </button>
            </Tooltip>
          </React.Fragment>
        }
        type="text"
        value=""
      />
    </styled.div>
  </styled.div>
</styled.div>
`;
